File: steelechallenges/trees/Binary Search Tree/challenge.js
class Node {
	constructor(value) {
		this.value = value;
		this.left = null;
		this.right = null;
	}
}

class BinarySearchTree {
	constructor() {
		this.root = null;
	}
	//helper recursive method for insertion
	compareAndInsert(currentNode, newNode) {
		//this implementation doesn't allow for duplicates
		if (currentNode.value === newNode.value) {
			console.log("Node already exists!", currentNode);
			return false;
		}
		//if node is greater than current
		if (newNode.value > currentNode.value) {
			//if there is no node assigned to right, assign it and break our of the loop
			if (!currentNode.right) {
				currentNode.right = newNode;
				return true;
			}
			//calls function recursively with new node for comparison
			this.compareAndInsert(currentNode.right, newNode);

			//if node is less than current
		} else if (newNode.value < currentNode.value) {
			// if there is no left node, assign it and break out of loop
			if (!currentNode.left) {
				currentNode.left = newNode;
				return true;
			}
			//recursively call function with new node for comparison
			this.compareAndInsert(currentNode.left, newNode);
		}

		return true;
	}
	//checks for root node, otherwise calls recursive function
	insert(value) {
		const newNode = new Node(value);
		if (!this.root) {
			this.root = newNode;
		} else {
			this.compareAndInsert(this.root, newNode);
		}
		return this;
	}
	compareAndFind(currentNode, queriedValue) {
		//node has been found
		if (currentNode.value === queriedValue) {
			return currentNode;
		}

		//if node is greater than current
		if (queriedValue > currentNode.value) {
			//if there is a node assigned to the right, recursively call with new values and return the value
			if (currentNode.right) {
				return this.compareAndFind(currentNode.right, queriedValue);
			}
			//we cannot search further, return false
			return false;
			//if node is less than current
		} else if (queriedValue < currentNode.value) {
			// if there is a node assigned to the left, recursively call with new values and return the value
			if (currentNode.left) {
				return this.compareAndFind(currentNode.left, queriedValue);
			}
			//we cannot search further, return false
			return false;
		}
	}

	find(value) {
		if (!this.root) {
			return false;
		}
		return this.compareAndFind(this.root, value);
	}

	printTree() {
		let currentNode = this.root;
		console.log(currentNode);
		while (currentNode.left) {
			this.printNode(currentNode.left);
			currentNode = currentNode.left;
		}
		currentNode = this.root;
		while (currentNode.right) {
			this.printNode(currentNode.right);
			currentNode = currentNode.right;
		}
	}

	printNode(node) {
		console.log(node);
	}
}

const testTree = new BinarySearchTree();
testTree.root = new Node(10);
testTree.root.right = new Node(15);
testTree.root.left = new Node(8);
testTree.insert(2);
testTree.insert(20);
testTree.insert(4);
testTree.insert(12);

testTree.printTree();

console.log(testTree.find(43));
console.log(testTree.find(20));
console.log(testTree.find(10));
console.log(testTree.find(2));

///Breadth First Search

File: steelechallenges/trees/Binary Search Tree/challenge.rb
class Node
    attr_accessor :value, :left, :right
    def initialize(value)
        self.value = value 
        self.left = nil
        self.right = nil
    end
end

class Binary_Search_Tree
    attr_accessor :root
    def initialize
        self.root = nil
    end

    def compare_and_insert(current_node, new_node)
        if(current_node.value == new_node.value)
          puts "Node already exists!", current_node
          return false;
        end
        if(new_node.value > current_node.value)
            if(!current_node.right)
              current_node.right = new_node
              return true
            end
            self.compare_and_insert(current_node.right, new_node)
        elsif(new_node.value < current_node.value)
            if(!current_node.left)
                current_node.left = new_node
                return true
            end
            self.compare_and_insert(current_node.left, new_node)
        end
        return true
    end

    def insert(value)
        new_node = Node.new(value)
        if(!self.root)
            self.root = new_node
        else 
            self.compare_and_insert(self.root, new_node)
        end
        return self
    end

    def compare_and_find(current_node, queried_value)
        if(current_node.value == queried_value)
            return current_node
        end

        if(queried_value > current_node.value)

            if(current_node.right)
                return self.compare_and_find(current_node.right, queried_value)
            end

            return false

        elsif(queried_value < current_node.value)
            
            if(current_node.left)
                return self.compare_and_find(current_node.left, queried_value)
            end

            return false
        end
    end

    def find(value)
        if(!self.root)
            return false
        end

        return self.compare_and_find(self.root, value)
    end

    def print_tree
        current_node = self.root
        self.print_node(current_node)
        while current_node
            self.print_node(current_node)
            current_node = current_node.left
        end
        current_node = self.root
        while current_node
            self.print_node(current_node)
            current_node = current_node.right
        end
    end

    def print_node(node)
        puts "Node Value:#{node.value}, Left:#{node.left}, Right:#{node.right}"
    end
end

test_tree = Binary_Search_Tree.new()
test_tree.root = Node.new(10)
test_tree.root.left = Node.new(6)
test_tree.root.right = Node.new(15)

puts test_tree.insert(2)
puts test_tree.insert(12)
# test_tree.print_tree
test_tree.print_node(test_tree.find(12))